modules/core/shared/src/main/scala/gem/parser/Misc.scala

```diff
@@ -65,5 +65,27 @@ trait MiscParsers {
 
   } namedOpaque s"frac($n)"
 
+
+  /** Force and return all remaining input without consuming it. */
+  val force: Parser[String] =
+    get.flatMap { s =>
+      opt(ensure(s.length + 1)) flatMap {
```

Comment:
How is `opt(ensure(s.length + 1))` not always `None`?

Reply:
atto does incremental parsing, so it may not see all the input at once. `get` lets you peek at [the remainder of] the current chunk of input. When it runs out you get a `Partial` parse result that says it's ok so far but you need to feed more (or say you're done) before you know for sure whether it succeeds or fails.

If your parser need a specific minimum amount of input you can use `ensure` to force more chunks to be read if there's not enough left in the current chunk. This is a suspended computation so what you "forcing" means is that you end up with is a `Partial` that will continue to be `Partial` until you feed it enough data to satisfy `ensure` or establish that there's really not enough input.

Neither `get` nor `ensure` actually "consume" the input, so once you establish that you have sufficient input that matches whatever you're looking for you can then `take` or `advance` or whatever to move forward.

So tl;dr `force` just repeatedly asks for more input than we have in order to pull everything into the buffer.